plugin/path_checkouts.go

```diff
@@ -5,11 +5,320 @@ import (
+	return b.handleCheckIn(ctx, req, fieldData, true)
+}
+
+func (b *backend) handleCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData, overrideCheckInEnforcement bool) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	serviceAccountNamesRaw, serviceAccountNamesSent := fieldData.GetOk("service_account_names")
+	var serviceAccountNames []string
+	if serviceAccountNamesSent {
+		serviceAccountNames = serviceAccountNamesRaw.([]string)
+	}
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+
+	// If check-in enforcement is overridden or disabled at the set level, we should consider it disabled.
+	disableCheckInEnforcement := overrideCheckInEnforcement || set.DisableCheckInEnforcement
+
+	respData := map[string]interface{}{
+		"check_ins": make([]string, 0),
+	}
+
+	// Build and validate a list of service account names that we will be checking in.
+	if len(serviceAccountNames) == 0 {
+		// It's okay if the caller doesn't tell us which service accounts they
+		// want to check in as long as they only have one checked out.
+		// We'll assume that's the one they want to check in.
+		for _, setServiceAccount := range set.ServiceAccountNames {
+			checkOut, err := b.checkOutHandler.Status(ctx, req.Storage, setServiceAccount)
+			if err != nil {
+				return nil, err
+			}
+			if checkOut.IsAvailable {
+				continue
+			}
+			if !canCheckIn(req, checkOut, disableCheckInEnforcement) {
+				continue
+			}
+			serviceAccountNames = append(serviceAccountNames, setServiceAccount)
+		}
+		switch len(serviceAccountNames) {
+		case 0:
+			// Everything's already currently checked in, nothing else to do here.
+			return &logical.Response{
+				Data: respData,
+			}, nil
+		case 1:
+			// This is what we need to continue with the check-in process.
+		default:
+			return logical.ErrorResponse(`when multiple service accounts are checked out, the "service_account_names" to check in must be provided`), nil
+		}
+	} else {
+		var cantCheckInErrs error
+		for _, serviceAccountName := range serviceAccountNames {
+			checkOut, err := b.checkOutHandler.Status(ctx, req.Storage, serviceAccountName)
+			if err != nil {
+				return nil, err
+			}
+			if !checkOut.IsAvailable {
+				continue
+			}
+			if !canCheckIn(req, checkOut, disableCheckInEnforcement) {
+				cantCheckInErrs = multierror.Append(cantCheckInErrs, fmt.Errorf(`"%s" can't be checked in because it wasn't checked out by the caller, please remove it and try again`, serviceAccountName))
+			}
+		}
+		if cantCheckInErrs != nil {
+			return logical.ErrorResponse(cantCheckInErrs.Error()), nil
+		}
+	}
+
+	respData["check_ins"] = serviceAccountNames
+	for _, serviceAccountName := range serviceAccountNames {
+		if err := b.checkOutHandler.CheckIn(ctx, req.Storage, serviceAccountName); err != nil {
+			return nil, err
+		}
```

Comment:
If we have multiple service accounts to check-in, should we attempt all of them and collect any errors instead? Here it seems if we have 5 to check-in and we fail on the first, we wont even attempt the other 4. 

That's assuming I have the right context here, and that a situation with multiple service accounts to check in is something that could happen.

Reply:
I suppose we could. If I checked them in and still returned errors for the remaining as a 4XX, then the client wouldn't retry. Yes, I'll do that, it would be a better UX.